byzcoin/proto.go

```diff
@@ -301,6 +301,45 @@ type StreamingResponse struct {
 	Block *skipchain.SkipBlock
 }
 
+// PaginateRequest is a request to get NumPages times the consecutive list of
+// PageSize blocks.
+type PaginateRequest struct {
+	// The first block to fetch
+	StartID skipchain.SkipBlockID
+	// Determines the length of the Blocks attribute in the  PaginateResponse.
+	// The list contains PageSize consecutive blocks
+	PageSize uint64
+	// The number of (asynchrounous) requests the service will return to the
+	// client. Requests are send in a consecutive order wrt their list of blocks
+	NumPages uint64
+	// If true then blocks are consecutive in the reverse order, ie. following
+	// backward links.
+	Backward bool
+	// Identifies the service's channel of a previously established connection.
+	StreamID []byte
+}
+
+// PaginateResponse is a reponse from a PaginateRequest. It returns the streamID
+// so that a new PaginateRequest can use an already created channel.
+type PaginateResponse struct {
+	// A list of consecutive blocks
+	Blocks []*skipchain.SkipBlock
+	// The page number index: relevant if the clients asked for more than one
+	// asynchrounous reply from the service.
+	PageNumber uint64
+	// Identifies the channel the service uses to send back messages. The client
+	// must provide it if he wants to send additional requests using the same
+	// channel.
+	StreamID []byte
+	// Tells if the result contains consecutive blocks in a reversed order.
+	Backward bool
+	// Used to tell the client if an error occured. Any error code not equal to
+	// 0 means that something special happened.
+	ErrorCode uint64
+	// A list of error messages in case something special happended.
```

Comment:
s/happended/happened/

"Something special" is kind of strange. Anyway, I don't see a place where more than one is returned. What is this for? What am I missing here?

Reply:
I actually forgot to make the final update that would use different error codes. Its added now.

One use case is when the user asks for too many blocks than the chain has. In that case, the service returns no blocks (its either all the requested blocks in a page, or none). But the user can check the error code, 5 in this case, and know that the list of error `errorText` contains the index of the last block, which the client can then use to make a new request with the correct number of blocks.